src/GoogleDriveAdapter.php

```diff
@@ -1086,117 +1086,164 @@ protected function createDirectory($name, $parentId)
+             * In this scenario, we send the entire file off as a string to have the mime type detected. Unfortunately, this leads to the entire
+             * file being loaded into memory again, separately from the copy we're holding.
+             */
+            $chunkSizeBytes = max(262144, min($chunkSizeBytes, floor($availableMemory / 262144) * 262144 / 4));
         }
 
-        $isResource = false;
-        if (is_resource($contents)) {
-            $fstat = @fstat($contents);
-            if (! empty($fstat['size'])) {
-                $isResource = true;
-            }
-            if (! $isResource) {
-                $contents = stream_get_contents($contents);
-            }
+        return (int)$chunkSizeBytes;
+    }
+
+    /**
+     * Normalise a Drive File that has been created
+     *
+     * @param Google_Service_Drive_DriveFile $uploadedFile
+     * @param string $localPath
+     * @param string $visibility
+     * @return array|bool
+     */
+    protected function normaliseUploadedFile($uploadedFile, $localPath, $visibility)
+    {
+        list ($parentId, $fileName) = $this->splitPath($localPath);
+
+        if (!($uploadedFile instanceof Google_Service_Drive_DriveFile)) {
+            return false;
         }
 
-        if ($isResource) {
-            // set chunk size (max: 100MB)
-            $chunkSizeBytes = 100 * 1024 * 1024;
-            $memory = $this->getIniBytes('memory_limit');
-            if ($memory > 0) {
-                $chunkSizeBytes = max(262144 , min([
-                    $chunkSizeBytes,
-                    (intval($memory / 4 / 256) * 256)
-                ]));
-            }
-            if ($fstat['size'] < $chunkSizeBytes) {
-                $isResource = false;
-                $contents = stream_get_contents($contents);
-            }
+        $this->cacheFileObjects[$uploadedFile->getId()] = $uploadedFile;
+        if (! $this->getFileObject($localPath)) {
+            $this->cacheFileObjectsByName[$parentId . '/' . $fileName] = $uploadedFile;
+        }
+        $result = $this->normaliseObject($uploadedFile, Util::dirname($localPath));
+
+        if ($visibility && $this->setVisibility($localPath, $visibility)) {
+            $result['visibility'] = $visibility;
         }
 
+        return $result;
+    }
+
+    /**
+     * Upload a PHP resource stream to Google Drive
+     *
+     * @param resource $resource
+     * @param string $localPath
+     * @param string $mime
+     * @return bool|Google_Service_Drive_DriveFile
+     */
+    protected function uploadResourceToGoogleDrive($resource, $localPath, $mime)
+    {
+        $fileSize = $this->getFileSizeBytes($resource);
+        $chunkSizeBytes = $this->detectChunkSizeBytes();
+
+        if ($fileSize <= $chunkSizeBytes) {
+            // If the resource fits in a single chunk, we'll just upload it in a single request
+            return $this->uploadStringToGoogleDrive($localPath, $mime, stream_get_contents($resource));
```

Comment:
@DynamoGeek Thank you for your contribution. It seems that the order of the arguments in this part is wrong.

Reply:
That's an embarrassing oversight that snuck in at the end. 😬

Thanks for the thorough review! 🥇